streams/test-utils/src/test/java/org/apache/kafka/streams/test/TopologyTestDriverTest.java

```diff
@@ -0,0 +1,573 @@
+
+    private final ConsumerRecordFactory<byte[], byte[]> consumerRecordFactory = new ConsumerRecordFactory<>(
+        new ByteArraySerializer(),
+        new ByteArraySerializer());
+
+    private final byte[] key1 = new byte[0];
+    private final byte[] value1 = new byte[0];
+    private final long timestamp1 = 42L;
+    private final ConsumerRecord<byte[], byte[]> consumerRecord1 = consumerRecordFactory.create(SOURCE_TOPIC_1, key1, value1, timestamp1);
+
+    private final byte[] key2 = new byte[0];
+    private final byte[] value2 = new byte[0];
+    private final long timestamp2 = 43L;
+    private final ConsumerRecord<byte[], byte[]> consumerRecord2 = consumerRecordFactory.create(SOURCE_TOPIC_2, key2, value2, timestamp2);
+
+    private TopologyTestDriver testDriver;
+    private final Properties config = new Properties() {
+        {
+            put(StreamsConfig.APPLICATION_ID_CONFIG, "test-TopologyTestDriver");
+            put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "dummy:1234");
+            put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+        }
+    };
+
+    private final static class Record {
+        private Object key;
+        private Object value;
+        private long timestamp;
+        private long offset;
+        private String topic;
+
+        Record(final ConsumerRecord consumerRecord) {
+            key = consumerRecord.key();
+            value = consumerRecord.value();
+            timestamp = consumerRecord.timestamp();
+            offset = consumerRecord.offset();
+            topic = consumerRecord.topic();
+        }
+
+        Record(final Object key,
+               final Object value,
+               final long timestamp,
+               final long offset,
+               final String topic) {
+            this.key = key;
+            this.value = value;
+            this.timestamp = timestamp;
+            this.offset = offset;
+            this.topic = topic;
+        }
+
+        @Override
+        public String toString() {
+            return "key: " + key + ", value: " + value + ", timestamp: " + timestamp + ", offset: " + offset + ", topic: " + topic;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final Record record = (Record) o;
+            return timestamp == record.timestamp &&
+                offset == record.offset &&
+                Objects.equals(key, record.key) &&
+                Objects.equals(value, record.value) &&
+                Objects.equals(topic, record.topic);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(key, value, timestamp, offset, topic);
+        }
+    }
+
+    private final static class Punctuation {
```

Comment:
Please see my other comment for `MockProcessor`: it's better to just reuse the existing one with improved punctuation configuration.

Reply:
Are you referring to the classes in test package? I think we should not depend on the test package, as test package is not public API and the new artifact should not depend on the test artifact from my point of view. We can refactor this later though, adding Mocks to this new test artifact and remove the mocks from our unit test to use the mocks from the public test artifact.

Reply:
That makes sense.